File: Programmers/Algorithm_Kit/01_Hash/005.cpp
#include <string>
#include <vector>
#include <iostream>
#include <map>
using namespace std;

vector<int> solution(vector<string> genres, vector<int> plays) {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);

    vector<int> answer;

    map<string, int> genres_plays; //장르(key) - 재생 횟수(value)
    map<string, map<int, int>> genres_idx_plays; //장르(key) - {노래 번호(key') - 재생 횟수(value')}(value)

    for (int i = 0; i < genres.size(); i++) {
        genres_plays[genres[i]] += plays[i]; //장르별 재생 횟수 저장
        genres_idx_plays[genres[i]][i] = plays[i]; //장르, 노래 번호별 재생 횟수 저장
    }

    while (genres_plays.size() > 0) {
        string genre = ""; //최대 재생 횟수를 가지는 장르
        int play = 0; //재생 횟수

        //plays를 통해 장르 찾기
        for (pair<string, int> pi : genres_plays) {
            if (play < pi.second) {
                play = pi.second;
                genre = pi.first;
            }
        }

        //결정된 장르 내 2곡씩 선정
        for (int i = 0; i < 2; i++) {
            int max_play = 0; //동일 장르 노래들 중 최대 재생 횟수
            int max_idx = -1; //해당 재생 횟수를 가지는 노래 번호

            //max_play를 통해 노래 번호 찾기
            for (pair<int, int> pi : genres_idx_plays[genre]) {
                if (max_play < pi.second) {
                    max_play = pi.second;
                    max_idx = pi.first;
                }
            }

            //장르 내 노래가 없다면 break
            if (max_idx == -1) {
                break;
            }

            answer.push_back(max_idx); //노래 번호 저장
            genres_idx_plays[genre].erase(max_idx); //저장된 노래 번호는 genres_idx_plays에서 삭제
        }

        genres_plays.erase(genre); //저장된 장르는 genres_plays에서 삭제
    }

    return answer;
}

File: programmers/algorithmkit/01_hash/005.java
import java.util.*;

class Solution {
    public List<Integer> solution(String[] genres, int[] plays) {
        Map<String, Integer> genresPlays = new HashMap<>();
        Map<String, Map<Integer, Integer>> genresIdxPlays = new TreeMap<>(); // 번호 낮은 순 정렬을 위해

        for (int i = 0; i < genres.length; i++) {
            genresPlays.put(genres[i], genresPlays.getOrDefault(genres[i], 0) + plays[i]); // 장르별 총 재생 횟수 저장

            Map<Integer, Integer> map = genresIdxPlays.get(genres[i]);

            if (map == null) {
                map = new TreeMap<>();
                genresIdxPlays.put(genres[i], map); // 장르별로 노래 번호별 재생 횟수 저장을 위한 map 생성
            }

            map.put(i, plays[i]);
        }

        List<Integer> answer = new ArrayList<>();

        while (!genresPlays.isEmpty()) {
            String genre = "";
            int play = 0;

            // 가장 많이 재생된 장르 찾기
            for (Map.Entry<String, Integer> entry : genresPlays.entrySet()) {
                if (play < entry.getValue()) {
                    play = entry.getValue();
                    genre = entry.getKey();
                }
            }

            // 해당 장르에서 최대 2곡 선택
            for (int i = 0; i < 2; i++) {
                int maxPlay = 0;
                int maxIdx = -1;

                // 해당 장르에서 가장 많이 재생된 노래 찾기
                for (Map.Entry<Integer, Integer> entry : genresIdxPlays.get(genre).entrySet()) {
                    if (maxPlay < entry.getValue()) {
                        maxPlay = entry.getValue();
                        maxIdx = entry.getKey();
                    }
                }

                // 장르 내 노래가 없다면 종료
                if (maxIdx == -1) {
                    break;
                }

                answer.add(maxIdx);

                // 선택한 노래 삭제
                genresIdxPlays.get(genre).remove(maxIdx);
            }

            // 처리한 장르 삭제
            genresPlays.remove(genre);
        }

        return answer;
    }
}

// map.get(key), map.remove(key), map.contains(key) : idx가 아닌 key를 통해 value를 찾는다